File: experiments/control-run/python/files.py
"""Finding `*.log` files directly inside a directory and reading them line by line, never outside it."""

import os
import stat
from collections.abc import Iterator
from pathlib import Path


class OutsideDirectory(ValueError):
    """A path that is not a regular file directly inside the root directory."""


def resolve_directory(directory: str) -> Path:
    """Resolve the directory once; raise NotADirectoryError or OSError if it is not one."""
    root = Path(directory).resolve(strict=True)
    if not root.is_dir():
        raise NotADirectoryError(f"{directory} is not a directory")
    return root


def list_log_files(root: Path) -> list[Path]:
    """Every regular, non-symlink `*.log` file directly inside root, in name order."""
    with os.scandir(root) as entries:
        names = [e.name for e in entries if e.name.endswith(".log") and e.is_file(follow_symlinks=False)]
    return [root / name for name in sorted(names)]


def check_inside(root: Path, path: Path) -> None:
    if path.parent != root or path.name in ("", ".", "..") or os.sep in path.name:
        raise OutsideDirectory(f"{path} is not directly inside {root}")
    if path.resolve(strict=True).parent != root:
        raise OutsideDirectory(f"{path} resolves outside {root}")


def read_lines(root: Path, path: Path) -> Iterator[bytes]:
    """Yield raw lines of one file, streamed; refuses symlinks, non-regular files and anything outside root."""
    check_inside(root, path)
    descriptor = os.open(path, os.O_RDONLY | os.O_NOFOLLOW)
    with open(descriptor, "rb") as handle:
        if not stat.S_ISREG(os.fstat(handle.fileno()).st_mode):
            raise OutsideDirectory(f"{path} is not a regular file")
        yield from handle

File: experiments/control-run/go/files.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"os"
	"slices"
	"strings"
)

// maxLine is the longest line kept, newline included; longer lines are malformed.
const maxLine = 64 * 1024

// ErrNoLogs reports that <dir> holds no *.log file.
var ErrNoLogs = errors.New("no .log file found")

// LogFiles lists the regular files named *.log directly inside root, in
// byte order of name. Symlinks and directories are skipped, so nothing
// outside root is ever listed; root.Open would refuse an escape regardless.
func LogFiles(root *os.Root) ([]string, error) {
	dir, err := root.Open(".")
	if err != nil {
		return nil, fmt.Errorf("cannot open directory: %w", err)
	}
	entries, readErr := dir.ReadDir(-1)
	if err := errors.Join(readErr, dir.Close()); err != nil {
		return nil, fmt.Errorf("cannot list directory: %w", err)
	}
	var names []string
	for _, e := range entries {
		if e.Type().IsRegular() && strings.HasSuffix(e.Name(), ".log") {
			names = append(names, e.Name())
		}
	}
	slices.Sort(names)
	return names, nil
}

// ScanFile feeds every line of the named file inside root to acc, holding
// at most one line in memory.
func ScanFile(root *os.Root, name string, acc *Accumulator) (err error) {
	f, err := root.Open(name)
	if err != nil {
		return fmt.Errorf("cannot open %q: %w", name, err)
	}
	defer func() {
		if closeErr := f.Close(); closeErr != nil && err == nil {
			err = fmt.Errorf("cannot close %q: %w", name, closeErr)
		}
	}()
	if err := scanLines(f, acc); err != nil {
		return fmt.Errorf("cannot read %q: %w", name, err)
	}
	return nil
}

// scanLines splits r on '\n'. A final line without a newline still counts;
// a line longer than maxLine counts once as malformed and is skipped.
func scanLines(r io.Reader, acc *Accumulator) error {
	br := bufio.NewReaderSize(r, maxLine)
	for {
		chunk, err := br.ReadSlice('\n')
		switch {
		case err == nil:
			acc.AddLine(string(chunk[:len(chunk)-1]))
		case errors.Is(err, bufio.ErrBufferFull):
			acc.AddMalformed()
			done, skipErr := skipLine(br)
			if skipErr != nil || done {
				return skipErr
			}
		case errors.Is(err, io.EOF):
			if len(chunk) > 0 {
				acc.AddLine(string(chunk))
			}
			return nil
		default:
			return err
		}
	}
}

// skipLine discards the rest of an over-long line; done reports end of input.
func skipLine(br *bufio.Reader) (done bool, err error) {
	for {
		_, err := br.ReadSlice('\n')
		switch {
		case err == nil:
			return false, nil
		case errors.Is(err, bufio.ErrBufferFull):
			continue
		case errors.Is(err, io.EOF):
			return true, nil
		default:
			return true, err
		}
	}
}
